server/src/main/java/io/druid/client/HttpServerInventoryView.java

```diff
@@ -0,0 +1,719 @@
+import org.joda.time.Duration;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * This class uses CuratorInventoryManager to listen for queryable server membership which serve segments(e.g. Historicals).
+ * For each queryable server, it uses HTTP GET /druid-internal/v1/segments (see docs in SegmentListerResource.getSegments(..).
+ */
+public class HttpServerInventoryView implements ServerInventoryView, FilteredServerInventoryView
+{
+  private final EmittingLogger log = new EmittingLogger(HttpServerInventoryView.class);
+  private final CuratorInventoryManager<DruidServer, Object> inventoryManager;
+  private final AtomicBoolean started = new AtomicBoolean(false);
+
+  private final ConcurrentMap<ServerCallback, Executor> serverCallbacks = new MapMaker().makeMap();
+  private final ConcurrentMap<SegmentCallback, Executor> segmentCallbacks = new MapMaker().makeMap();
+
+  private final ConcurrentMap<SegmentCallback, Predicate<Pair<DruidServerMetadata, DataSegment>>> segmentPredicates = new MapMaker()
+      .makeMap();
+  private final Predicate<Pair<DruidServerMetadata, DataSegment>> defaultFilter;
+  private volatile Predicate<Pair<DruidServerMetadata, DataSegment>> finalPredicate;
+
+  // For each queryable server, a name -> DruidServerHolder entry is kept
+  @GuardedBy("inventory")
+  private final Map<String, DruidServerHolder> servers = new HashMap<>();
+
+  private volatile ExecutorService executor;
+
+  // a queue of queryable server names for which worker threads in executor initiate the segment list call i.e.
+  // DruidServerHolder.updateSegmentsListAsync(..) which updates the segment list asynchronously and adds itself
+  // to this queue again for next update.
+  private final BlockingQueue<String> queue = new LinkedBlockingDeque<>();
+
+
+
+  private final HttpClient httpClient;
+  private final ObjectMapper smileMapper;
+  private final HttpServerInventoryViewConfig config;
+
+  @Inject
+  public HttpServerInventoryView(
+      final @Json ObjectMapper jsonMapper,
+      final @Smile ObjectMapper smileMapper,
+      final @Global HttpClient httpClient,
+      final CuratorFramework curator,
+      final String announcementsPath,
+      final Predicate<Pair<DruidServerMetadata, DataSegment>> defaultFilter,
```

Comment:
`Pair` is a horrible class to use in actual public APIs or any other API contract.  The reason is because it can quickly become difficult to know what the lhs and rhs are actually representing.  As a general code design rule, it's better to only use `Pair` as a convenience for internal things and if you are actually creating a contract to create a class that represents your pair of items with actual names.